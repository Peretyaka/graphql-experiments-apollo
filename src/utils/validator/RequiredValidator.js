import AbstractValidator from './AbstractValidator'

export default class RequiredValidator extends AbstractValidator {

  static defaults = {
    message: '',
  }

  /**
   * @param {string} value
   * @returns {bool}
   */
  validateValue(value) { // eslint-disable-line
    return !!value
  }
}
