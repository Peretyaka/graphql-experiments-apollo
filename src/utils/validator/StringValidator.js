import {isNumber, isString} from 'lodash'
import AbstractValidator from './AbstractValidator'

export default class StringValidator extends AbstractValidator {

  static defaults = {
    message: '',

    /**
     * Specifies the exact length limit of the value to be validated
     * @type {number}
     */
    length: null,

    /**
     * Specifies the minimum length of the value to be validated
     * @type {number}
     */
    min: null,

    /**
     * Specifies the maximum length limit of the value to be validated
     * @type {number}
     */
    max: null,
  }

  /**
   * @param {string} value
   * @returns {bool}
   */
  validateValue(value) {
    return this.checkString(value) &&
           this.checkLength(value) &&
           this.checkMax(value) &&
           this.checkMin(value)
  }

  checkString(value) { // eslint-disable-line
    return isString(value)
  }

  checkLength(value) {
    if (isNumber(this.options.length)) {
      return value.length === this.options.length
    }

    return true
  }

  checkMax(value) {
    if (isNumber(this.options.max)) {
      return value.length <= this.options.max
    }

    return true
  }

  checkMin(value) {
    if (isNumber(this.options.min)) {
      return value.length >= this.options.min
    }

    return true
  }
}
