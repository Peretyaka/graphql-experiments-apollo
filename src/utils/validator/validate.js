/**
 * @param {Object} data - object of fields that must be validated
 * @param {Object} rules
 *
 * Example of data:
 *
 * {
 *   foo: 'foo',
 *   bar: 'bar'
 * }
 *
 * Rules must be in such format, for e.g.:
 *
 * {
 *   foo: [
 *     new RequiredValidator({
 *       message: 'Foo should not be empty',
 *     })
 *   ],
 *   bar: [
 *     new RequiredValidator({
 *       message: 'Bar should not be empty',
 *     }),
 *     new StringValidator({
 *       message: `Bar must be at least 5 and not more than 100 characters`,
 *       min: 5,
 *       max: 100,
 *     })
 *   ]
 * }
 *
 * @returns {Promise}
 */
export default (data = {}, rules = {}) => new Promise((resolve, reject) => {
  const errors = {}
  const validators = []

  for (const field in rules) {
    rules[field].forEach((rule) => {
      // Add all rules to array for waiting for promise all (see below)
      validators.push(rule.validate(data[field])
        .catch((error) => {
          if (!errors[field]) {
            errors[field] = []
          }

          errors[field].push(error)
        })
      )
    })
  }

  Promise.all(validators).then(() => {
    // If errors have almost one key it mean that there are errors
    if (Object.keys(errors).length) {
      reject(errors, data)
    }

    resolve(data)
  })
})
