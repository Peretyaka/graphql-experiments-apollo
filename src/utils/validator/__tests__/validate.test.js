import validate from '../validate'
import RequiredValidator from '../RequiredValidator'
import StringValidator from '../StringValidator'

describe('Validate', () => {
  const errorMessage = 'ERROR'

  it('Should return original data on success validation', (done) => {
    const inputData = {
      foo: 'Foo',
      bar: 'Bar',
    }
    const rules = {
      foo: [new RequiredValidator({message: errorMessage})],
      bar: [
        new RequiredValidator({message: errorMessage}),
        new StringValidator({message: errorMessage, length: 3}),
      ],
    }

    validate(inputData, rules)
      .then((data) => {
        expect(data).toEqual(inputData)
        done()
      })
      .catch((error) => done.fail(error))
  })

  it('Should return errors in correct format', (done) => {
    const inputData = {
      foo: '',
      bar: '',
    }
    const rules = {
      foo: [new RequiredValidator({message: `${errorMessage}1`})],
      bar: [
        new RequiredValidator({message: `${errorMessage}2`}),
        new StringValidator({message: `${errorMessage}3`, length: 3}),
      ],
    }

    validate(inputData, rules)
      .then(() => done.fail())
      .catch((errors) => {
        expect(errors).toMatchSnapshot()
        done()
      })
  })

  it('Should avoid validation on fields with no rules specified on them', (done) => {
    const inputData = {
      foo: '',
      bar: 'Bar',
    }
    const rules = {
      foo: [new RequiredValidator({message: errorMessage})],
    }

    validate(inputData, rules)
      .then(() => done.fail())
      .catch((errors) => {
        expect(errors).toEqual({foo: [errorMessage]})
        done()
      })
  })
})
