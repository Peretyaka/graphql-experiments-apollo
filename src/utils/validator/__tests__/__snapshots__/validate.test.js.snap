// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Validate Should return errors in correct format 1`] = `
Object {
  "bar": Array [
    "ERROR2",
    "ERROR3",
  ],
  "foo": Array [
    "ERROR1",
  ],
}
`;
